
#ifndef __GraspPlanner_WINDOW_H_
#define __GraspPlanner_WINDOW_H_

#include <VirtualRobot/VirtualRobot.h>
#include <VirtualRobot/Robot.h>
#include <VirtualRobot/VirtualRobotException.h>
#include <VirtualRobot/Nodes/RobotNode.h>
#include <VirtualRobot/XML/SceneIO.h>
#include <VirtualRobot/Visualization/VisualizationFactory.h>
#include <VirtualRobot/Visualization/CoinVisualization/CoinVisualization.h>
#include <VirtualRobot/Obstacle.h>
#include <VirtualRobot/ManipulationObject.h>
#include <VirtualRobot/Obstacle.h>

#include "GraspPlanning/GraspStudio.h"
#include "GraspPlanning/GraspQuality/GraspQualityMeasureWrenchSpace.h"
#include "GraspPlanning/GraspPlanner/GenericGraspPlanner.h"
#include "GraspPlanning/ApproachMovementSurfaceNormal.h"

#include <string.h>
#include <QtCore/QtGlobal>
#include <QtGui/QtGui>
#include <QtCore/QtCore>

#include <Inventor/sensors/SoTimerSensor.h>
#include <Inventor/nodes/SoEventCallback.h>
#include <Inventor/Qt/viewers/SoQtExaminerViewer.h>
#include <Inventor/Qt/SoQt.h>
#include <Inventor/nodes/SoSeparator.h>


#include <vector>

#include "ui_GraspPlanner.h"

class GraspPlannerWindow : public QMainWindow
{
	Q_OBJECT
public:
	GraspPlannerWindow(std::string &robotFile, std::string &eefName, std::string &preshape, std::string &objectFile, Qt::WFlags flags = 0);
	~GraspPlannerWindow();

	/*!< Executes the SoQt mainLoop. You need to call this in order to execute the application. */
	int main();

public slots:
	/*! Closes the window and exits SoQt runloop. */
	void quit();

	/*!< Overriding the close event, so we know when the window was closed by the user. */
	void closeEvent(QCloseEvent *event);

	void resetSceneryAll();


	void closeEEF();
	void openEEF();
	void colModel();
	void frictionConeVisu();
	void showGrasps();

	void buildVisu();

	void plan();
	void save();

protected:
    
    void loadRobot();
	void loadObject();
	
	void setupUI();

	static void timerCB(void * data, SoSensor * sensor);
	Ui::GraspPlanner UI;
	SoQtExaminerViewer *viewer; /*!< Viewer to display the 3D model of the robot and the environment. */
		
	SoSeparator *sceneSep;
	SoSeparator *robotSep;
	SoSeparator *objectSep;
	SoSeparator *frictionConeSep;
	SoSeparator *graspsSep;

	VirtualRobot::RobotPtr robot;
	VirtualRobot::RobotPtr eefCloned;
	VirtualRobot::ObstaclePtr object;
	VirtualRobot::EndEffectorPtr eef;
	    
	VirtualRobot::GraspSetPtr grasps;


	VirtualRobot::EndEffector::ContactInfoVector contacts;

	
	std::string robotFile;
	std::string objectFile;
	std::string eefName;
	std::string preshape;

	SoSeparator *eefVisu;

	GraspStudio::GraspQualityMeasureWrenchSpacePtr qualityMeasure;
	GraspStudio::ApproachMovementSurfaceNormalPtr approach;
	GraspStudio::GenericGraspPlannerPtr planner;

    boost::shared_ptr<VirtualRobot::CoinVisualization> visualizationRobot;
    boost::shared_ptr<VirtualRobot::CoinVisualization> visualizationObject;
};

#endif // __GraspPlanner_WINDOW_H_
